js/lib/hooks.js

```diff
@@ -36,92 +106,203 @@ elgg.register_hook_handler = function(name, type, handler, priority) {
-	}
+	var new_list = [],
+		ret = false;
+	$.each(hooks[name][type], function(i, registration) {
+		if (registration.handler === handler) {
+			ret = true;
+		} else {
+			new_list.push(registration);
+		}
+	});
 
-	return hooks[name][type].insert(handler, priority);
+	hooks[name][type] = new_list;
+	return ret;
 };
 
-/**
- * Emits a hook.
- *
- * Loops through all registered hooks and calls the handler functions in order.
- * Every handler function will always be called, regardless of the return value.
- *
- * @warning Handlers take the same 4 arguments in the same order as when calling this function.
- * This is different from the PHP version!
- *
- * @note Instant hooks do not support params or values.
- *
- * Hooks are called in this order:
- *	specifically registered (event_name and event_type match)
- *	all names, specific type
- *	specific name, all types
- *	all names, all types
- *
- * @param {String} name   Name of the hook to emit
- * @param {String} type   Type of the hook to emit
- * @param {Object} params Optional parameters to pass to the handlers
- * @param {Object} value  Initial value of the return. Can be mangled by handlers
- *
- * @return {Boolean}
- */
-elgg.trigger_hook = function(name, type, params, value) {
-	elgg.assertTypeOf('string', name);
-	elgg.assertTypeOf('string', type);
+!function(){
+
+	/**
+	 * Emits a synchronous hook, calling only synchronous handlers
+	 *
+	 * Loops through all registered hooks and calls the handler functions in order.
+	 * Every handler function will always be called, regardless of the return value.
+	 *
+	 * @warning Handlers take the same 4 arguments in the same order as when calling this function.
+	 * This is different from the PHP version!
+	 *
+	 * @note Instant hooks do not support params or values.
+	 *
+	 * Hooks are called in priority order.
+	 *
+	 * @param {String} name   The hook name.
+	 * @param {String} type   The hook type.
+	 * @param {Object} params Optional parameters to pass to the handlers
+	 * @param {Object} value  Initial value of the return. Can be modified by handlers
+	 *
+	 * @return {*}
+	 */
+	elgg.trigger_hook = function(name, type, params, value) {
+		return trigger(name, type, params, value, false);
+	};
+
+	/**
+	 * Emits an asynchronous hook.
+	 *
+	 * Loops through all registered hooks and calls the handler functions in order.
+	 * Every handler function will always be called, regardless of the return value.
+	 *
+	 * @param {String} name   The hook name.
+	 * @param {String} type   The hook type.
+	 * @param {Object} params Optional parameters to pass to the handlers
+	 * @param {Object} value  Initial value of the return. Can be modified by handlers
+	 *
+	 * @return {Promise}
+	 */
+	elgg.trigger_async_hook = function(name, type, params, value) {
+		return trigger(name, type, params, value, true);
+	};
+
+	/**
+	 * Time limit per async handler
+	 *
+	 * @type {Number}
+	 */
+	elgg.config.async_handler_timeout = 1000 * 20;
```

Comment:
For 99% of things it's probably plenty long, but I don't see any reason to artificially restrict it.  I suppose there could be edge cases where it could be longer, but those would be pretty rare.  Is there a possibility for the handler to extend it if there is reason to expect it might legitimately need longer?